### PR TITLE
Ensure that onError with timeout is called after subscription cancel

### PR DESCRIPTION
**Pull request by bondolo**
Motivation:
In some cases the `cancel()` which resulted from a timeout did not
result in the `onError()` also being called with the timeout exception
because the cancel operation changed the subscriber state.

Modifications:
If the operator had not completed at the time of the timeout then the
`onError()` should be called regardless of the state changes which
happen during `cancel()`.

Result:
The reason for the cancel is exposed via the `onError()`.

### CODE REVIEW

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutSingle.java
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutSingle.java:194]: The `subscriberState` changes only in 2 places: here and inside `onSubscribe`. How is it possible that `cancel()` modifies the state?

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutPublisher.java
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutPublisher.java:293]: Why `finally` is required here? For cases when `onSubscribe/cancel` throws?

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutSingle.java
Comment (bondolo) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutSingle.java:194]: That part was possibly not essential. The main motivation was for the `cancel()` throws case and specifically when it is throwing the channel closed exception and we want the timeout exception "true cause" as the onError result.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutPublisher.java
Comment (bondolo) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutPublisher.java:293]: Yes, to ensure that the original "true cause" is not lost.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutSingle.java
Comment (idelpivnitskiy) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutSingle.java:194]: Interesting, the RS spec says that `cancel()` must [return normally](https://github.com/reactive-streams/reactive-streams-jvm#3.15). If you have seen that `cancel()` may throw, we should investigate the cause and fix it. Can you please provide more context when it happens? Do you have a reproducer?

Comment (bondolo) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutSingle.java:194]: I think I have a better explanation now. It is not that cancel() might throw an exception. You are correct, it should never throw. The problem this addresses is that the call to cancel() might cause an onError() propagation and this can only be done once. By updating the subscriber state before calling cancel() the updater reserves the right for themselves to call the onError() with the cause they intended rather than someone else in the cancel() chain calling it with their cause.

In the particular case of timeout the exposed cause via onError() was showing up as StacklessClosedChannelException because the channel threw that when canceled. By updating the subscription state before calling cancel() we are suppressing any onError() causes being propagated up from within the cancel(). When the cancel() completes we call onError() with the cause we want to expose, TimeoutException.

Comment (bondolo) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutSingle.java:194]: In a perfect world any onError() which occurs while cancel() is being called would be captured as a suppressed exception for the TimeoutException.

**Review by Scottmitch — APPROVED** [review]
we will need similar changes for `Publisher` to make sure the timeout exception is propagated but we can do that in a followup PR as the state machine is more involved.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutSingle.java
Comment (idelpivnitskiy) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutSingle.java:194]: I don't want to block this PR, feel free to merge if it's already approved. I will appreciate if someone can help me to understand what I'm missing, but we can do that after the PR is merged or offline.

Thanks for the detailed explanation. This is approximately what I thought originally after reading the commit description. Hence my question: 

> How is it possible that `cancel()` modifies the state?

`target.onError(t)` is protected by two states: `subscriberStateUpdater` and `cancellableUpdater`. If we are inside the current `if` statement of the `run()` method, we flip the cancellable to `LOCAL_IGNORE_CANCEL` at the beginning. As the result, if the `oldCancellable.cancel()` generates `onError`, it will be ignored because the `onError` also checks the current `cancellable` and propagates it only if it's not `LOCAL_IGNORE_CANCEL`:
```java
        @Override
        public void onError(final Throwable t) {
            if (cancellableUpdater.getAndSet(this, LOCAL_IGNORE_CANCEL) != LOCAL_IGNORE_CANCEL) {
                try {
                    stopTimer();
                } finally {
                    target.onError(t);
                }
            }
        }
```

From what you describe it looks like `StacklessClosedChannelException` is correctly propagated because it's generated **before** the timeout expires (before `run()` executes).

Inside the `timeout-http` branch I reverted this commit and repeatedly run the `ResponseTimeoutTest`. Could not reproduce an issue.

Is it possible to write a repeatable reproducer for the problem this PR tries to fix?

**Review by Scottmitch — COMMENTED** [review]
@idelpivnitskiy brings up a good point, lets hold off on merging this until we have a reproducer which demonstrates the case this fixes.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutSingle.java
Comment (bondolo) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutSingle.java:194]: I am unable to reproduce the original problem as well.

**Comment by bondolo** [comment]
This change does not appear to be necessary. Closing without integration.

**Pull request closed without merging**
